File: csharp/diamond/Diamond.cs
using System.Text;

public static class Diamond
{
    public static string Make(char target)
    {
        var side = target - 'A' + 1;
        var rows = side * 2 - 1;

        var result = new List<string>();
        for (var r = 0; r < rows; r++)
        {
            var d = Math.Min(r, rows - 1 - r);
            var c = (char)(d + 'A');

            var outerSpaces = side - d - 1;

            var sb = new StringBuilder();

            sb.Append(' ', outerSpaces);
            sb.Append(c);

            if (d > 0)
            {
                sb.Append(' ', 2 * d - 1);
                sb.Append(c);
            }

            sb.Append(' ', outerSpaces);

            result.Add(sb.ToString());
        }

        return string.Join('\n', result);
    }
}

File: cpp/diamond/diamond.cpp
#include "diamond.h"
#include <algorithm>

namespace diamond
{

    std::vector<std::string> rows(char target)
    {
        int side = target - 'A' + 1;
        int rows = side * 2 - 1;

        std::vector<std::string> result{};

        for (int r = 0; r < rows; r++)
        {
            int d = std::min(r, rows - 1 - r);
            int c = (d + 'A');

            int outer_spaces = side - d - 1;

            std::string row{};

            row.append(outer_spaces, ' ');
            row += c;

            if (d > 0)
            {
                row.append(2 * d - 1, ' ');
                row += c;
            }

            row.append(outer_spaces, ' ');

            result.push_back(row);
        }

        return result;
    }

} // namespace diamond
